### ISSUE TITLE
disable certain packages or espanso as a whole when matching specific program

### ISSUE THREAD

**Issue by AryToNeX** [federico-terzi/espanso#177]
Hi there! Glad to see Italian people writing actually useful projects! :it: 
However, I have a really huge problem when using Telegram Desktop and the emoji-all package from hub. Thing is, Espanso gets triggered alongside Telegram's own 'emoji expander' when I type :heart:. Best case scenario, I only want to send a heart; worst case scenario, Espanso eats the six last chars of my message! I didn't find anything useful in the configuration files to prevent this behaviour, so I am asking you.

**Comment by federico-terzi** [comment]
Hey @AryToNeX,
Thank you for the kind words! You could automatically disable espanso when using Telegram, and it's pretty easy to do with [App-specific configurations](https://espanso.org/docs/configuration/#application-specific-configurations).

Which version of espanso are you using? As the recent v0.5.0 release changed names a bit. 
Here's how you can disable espanso when using telegram: 
Add a `telegram.yml` in the `espanso/user` folder with the following content:

```yaml
filter_title: "Telegram"
enable_active: false
```

and then restart espanso.

In case you are using a prior version, you should change `enable_active: false` with `disabled: true`, though I recommend you to update espanso as it ships with many new features.

Let me know if that works

Buona giornata 🇮🇹

**Comment by AryToNeX** [comment]
I copied and pasted that snippet to my app-specific configuration file and it works! However, there are some issues with warnings now...
```
Warning: trigger ':+1:' is conflicting with ':+1:' and may not behave as intended
Warning: trigger ':-1:' is conflicting with ':-1:' and may not behave as intended
Warning: trigger ':bee:' is conflicting with ':bee:' and may not behave as intended
...and so on for most of the emoji-all package bindings
```

Are matches defined twice for every app-specific configuration?

**Comment by federico-terzi** [comment]
Hey,
> I copied and pasted that snippet to my app-specific configuration file and it works! However, there are some issues with warnings now...
> 
> ```
> Warning: trigger ':+1:' is conflicting with ':+1:' and may not behave as intended
> Warning: trigger ':-1:' is conflicting with ':-1:' and may not behave as intended
> Warning: trigger ':bee:' is conflicting with ':bee:' and may not behave as intended
> ...and so on for most of the emoji-all package bindings
> ```

Thanks for the feedback, this is definitely a bug. It seems like I didn't consider this edge case when testing the new conflict warning feature, I'll ship a fix in the next release.

For now, just ignore the warnings, they don't do anything else other than complaining :)

> Are matches defined twice for every app-specific configuration?
> Also, I noticed that now I cannot even define app-specific matches and expect them to work since Espanso is fully disabled. Is there a way to exclude all matches but the app specific ones?

App-specific configs inherit all matches from the `default.yml` config by default. However, if you want to define some matches that work *only* on a specific app, without inheriting the default ones ( and consequently all the packages ), you can add the `exclude_default_entries: true` option:

```yaml
filter_title: "Telegram"
exclude_default_entries: true

matches:
  - trigger: "match"
    replace: "test"
```

**Comment by federico-terzi** [comment]
> I cannot see why we cannot override the `conflict_check` variable on a per-app basis though;

The conflict check is only used as a suggestion to debug misbehaving matches, it doesn't do anything other than warning the user. Therefore, I don't really see how disabling it on a per-application can bring much value to the user.

>  and as far as I am reading on the configuration dump, there's no way to disable packages on a per-app basis too, which is something I definitely would love to see on this software.

This could be a future feature, though I need to think carefully about the best way to implement it.

Thanks for your help :)

**Comment by LaurentFough** [comment]
@federico-terzi Is it possible to specify `filter_title` as regex.

My use case is: instead of creating multiple app-specific user files for the terminals that I use, that I create one app specific file and use regex for the filter_title, so that it recognizes the apps that I want.

**Comment by federico-terzi** [comment]
@LaurentFough yes you can :) In fact, all the `filter_*` properties are regexes

**Comment by dugite-code** [comment]
Question in regards to the regex are you able to use lookahead? I wanted to have a config that works everywhere except in AutoCAD whilst not disabling the default config.

I gave `^(?!Autodesk AutoCAD.*).*$` a try

**Comment by federico-terzi** [comment]
@dugite-code Unfortunately, the rust `Regex` crate does not support look-ahead and look-behind :(
